gaps/gap_node_descriptor/gap_node_descriptor.md

```diff
@@ -0,0 +1,109 @@
+
+#### Permissions
+
+Details of how the node is permitted to use the Golem network. The content of this property is governed by the [permissions schema](https://golem.network/schemas/v1/permissions.schema.json).
+
+#### Validity Period
+
+The time period for which this node descriptor is valid. Semantics defined in the [certificate schema](https://github.com/golemfactory/golem-architecture/blob/master/gaps/gap-25_golem_certificates/gap-25_golem_certificates.md#validity-period).
+
+#### Signature
+
+Semantics of the signature are defined in the [certificate schema](https://github.com/golemfactory/golem-architecture/blob/master/gaps/gap-25_golem_certificates/gap-25_golem_certificates.md#signature) with one difference. A node descriptor cannot be used to create signatures so it cannot be self signed.
+
+### Signature creation and verification
+
+The process is quite similar to what is described for [certificates](https://github.com/golemfactory/golem-architecture/blob/master/gaps/gap-25_golem_certificates/gap-25_golem_certificates.md#signature-creation-and-verification), but the signature is created for data in the `nodeDescriptor` property instead of the `certificate` one.
+
+#### Creating a signature
+
+In order to cryptographically sign a node descriptor the following steps need to be taken:
+
+1. Create a binary representation of the `nodeDescriptor` property by serializing the content via the `JSON Canonicalization Scheme`
+2. Sign the binary data with the chosen signature algorithm using the private key of the signer. The private key must be the one connected to the public key specified in the signer's certificate.
+3. Create the `signature` property in the JSON certificate and add the following details
+    - name of the hash and encryption algorithm (details of the signature algorithm) used to create the signature into the `algorithm` property
+    - binary data of the signature into `value` property
+    - the `signer` property of the `signature` must contain the signing certificate
+
+#### Verifying a signature
+
+1. Verify that the signing certificate contained in `signature.signer` is allowed to sign node descriptors (it's `keyUsage` property is either set to the string `all` or the list contains `signNode`).
+2. Create a binary representation of the `nodeDescriptor` property by serializing the content via the `JSON Canonicalization Scheme`.
+3. Using the signature algorithm matching the one used for creating the signature (`signature.algorithm.hash` and `signature.algorithm.encryption`) and the signer's public key (the `certificate.publicKey` property of the signing certificate), verify that the signature is valid for the binary data obtained in step 2.
+
+### Node descriptor in the Demand
+
+The node descriptor is added as a json object in the `golem.node.descriptor` property of the demand and becomes integral part of the demand during negotiation.
+
```

Comment:
OK, this is a very tricky part IMHO.
As far as I understand, the concept is for the descriptor to be signed with a private key of the node. That key is managed by the golem daemon on the node, and MUST NOT leave the daemon. In other words, the agent app is not able to sign the descriptor - the signing must be done by the daemon. 

One possible scenario of achieving this is:
- Agent app creates unsigned descriptor body and adds it to the `golem.node.descriptor` property.
- The Golem daemon recognizes `golem.node.descriptor`, validates its content and applies signature (how can Demand creator control the signing algorithm to be used?)
- The Demand with signed certificate is then sent to the Golem network for matching.

Reply:
There is some misconception here:
The node descriptor is not signed for each request. It is signed by a [Golem certificate](https://github.com/golemfactory/golem-architecture/pull/67) and have a validity period. The signing certificate transfers trust to use the features granted in the node descriptor to a node with the given node ID. I think it is very clear that the signer is a certificate from the schema and from the GAP in points:
#### Signature
### Signature creation and verification

The node ID is the public key of the node, so it is not feasible to spoof as without the private key the challenge-response verification of the node ID would fail during communication handshake. So a node's signed node descriptor cannot be used by another node, they would need to have the node's keypair.

Reply:
Fair enough, I obviously blindly assumed the signing is to be done by node's private key. By miss, question closed.